packages/zoe/src/contractFacet/vatRoot.js

```diff
@@ -8,51 +8,85 @@
 // `yarn build-zcfBundle`.
 
 import { Far } from '@endo/marshal';
+import { E } from '@endo/far';
 
 import '../../exported.js';
 import '../internal-types.js';
 
 import { makeZCFZygote } from './zcfZygote.js';
 
+const { details: X } = assert;
+
 /**
  * @param {VatPowers & { testJigSetter: TestJigSetter }} powers
- * @returns {{ executeContract: ExecuteContract}}
+ * @param {{contractBundleCap: BundleCap, zoeService: ZoeService, invitationIssuer: Issuer, privateArgs?: any}} vatParameters
+ * @param {import('@agoric/vat-data').Baggage} baggage
  */
-export function buildRootObject(powers) {
+export async function buildRootObject(powers, vatParameters, baggage) {
   // Currently, there is only one function, `executeContract` called
   // by the Zoe Service. However, when there is kernel support for
   // zygote vats (essentially freezing and then creating copies of
   // vats), `makeZCFZygote`, `zcfZygote.evaluateContract` and
   // `zcfZygote.startContract` should exposed separately.
   const { testJigSetter } = powers;
+  const { contractBundleCap } = vatParameters;
+  assert(
+    contractBundleCap,
+    X`expected vatParameters.contractBundleCap ${vatParameters}`,
+  );
+  let { zoeService, invitationIssuer } = vatParameters;
+  const firstTime = !baggage.has('DidStart');
+  if (firstTime) {
+    baggage.init('DidStart', 'DidStart');
+    baggage.init('zoeService', zoeService);
+    baggage.init('invitationIssuer', invitationIssuer);
+  } else {
+    assert(!zoeService, 'On restart zoeService must not be in vatParameters');
+    zoeService = baggage.get('zoeService');
+
+    assert(
+      !invitationIssuer,
+      'On restart invitationIssuer must not be in vatParameters',
+    );
+    invitationIssuer = baggage.get('invitationIssuer');
+  }
 
-  /** @type {ExecuteContract} */
-  const executeContract = (
-    bundleOrBundleCap,
+  // make zcfZygote with contract-general state and kinds initialized
+  const zcfZygote = await makeZCFZygote(
+    powers,
     zoeService,
     invitationIssuer,
-    zoeInstanceAdmin,
-    instanceRecordFromZoe,
-    issuerStorageFromZoe,
-    privateArgs = undefined,
-  ) => {
-    /** @type {ZCFZygote} */
-    const zcfZygote = makeZCFZygote(
-      powers,
-      zoeService,
-      invitationIssuer,
-      testJigSetter,
+    testJigSetter,
+    contractBundleCap,
+    baggage,
+  );
+
+  // snapshot zygote here //////////////////
+
+  if (!firstTime) {
+    return E.when(E(zcfZygote).restartContract(vatParameters.privateArgs), () =>
+      Far('upgraded contractRunner', {}),
     );
-    zcfZygote.evaluateContract(bundleOrBundleCap);
-    return zcfZygote.startContract(
+  }
+
+  return Far('contractRunner', {
+    // initialize instance-specific state of the contract
+    /** @type {StartZcf} */
+    startZcf: (
       zoeInstanceAdmin,
       instanceRecordFromZoe,
       issuerStorageFromZoe,
       privateArgs,
```

Comment:
This looks like a regression. This parameter is optional. The optionality is still expressed in the type so it should remain in the JS.

Reply:
done